### DISCUSSION TITLE
Rewrites by header doesn't work

### DISCUSSION THREAD

**Discussion by andriiradkevych** [vercel/next.js#37841] [Help]
**pages/_middleware.ts**

```
import { NextRequest, NextResponse } from 'next/server';

const isMobile = (userAgent: string) =>
  /iPhone|iPad|iPod|Android/i.test(userAgent);

const propName = 'x-rewrite';

enum Device {
  desktop = 'no',
  mobile = 'yes',
}

export function middleware(req: NextRequest) {
  const userAgent = req.headers.get('user-agent');

  const res = NextResponse.next();

  if (userAgent) {
    if (isMobile(userAgent)) {
      res.headers.set(propName, Device.mobile);
      console.log(res.headers, 'res');
      return res;
    } else {
      res.headers.set(propName, Device.desktop);
      return res;
    }
  }
}

```

**next.config.js**

```
  async rewrites() {
    return {
      beforeFiles: [
        {
          source: '/football/livescore',
          destination: '/football/livescore/mobile',
          has: [
            {
              type: 'header',
              key: 'x-rewrite',
              value: '(?<rewrite>yes|true)',
            },
          ],
        },
        {
          source: '/football/livescore/:league',
          destination: '/football/livescore/mobile/:league',
          has: [
            {
              type: 'header',
              key: 'x-rewrite',
              value: '(?<rewrite>yes|true)',
            },
          ],
        },
      ],
    };
  },
  ```
  
  
![image](https://user-images.githubusercontent.com/26725494/174583925-682c92d9-e6e4-4d3d-afb1-78fc23953f5c.png)

Above you can see my folder structure , you can see my `rewrites` function , and middleware. What I want to happens there , when request come I detect is it mobile device , then I add header property , and assign some value to it , after that I want my `rewrites` function change destination if it is mobile , but it doesn't work . I've also tried to make it with cookie , but it only work , by second reload of page . Could someone help , what I'm doing wrong ?

**Comment by icyJoseph** [comment]
HI,

I think the issue is that the middleware `request`, header modifiers do not arrive to the rest of the application. I've experienced this when trying to see if a user is logged in and forward that decision, or by trying to guard for certain headers. It didn't work. What worked for me was to set the header on the response object, and then I'd check that on the application instead.

For your use case, I'd ask, why not do the rewrites at the middleware already? You can even write tests for them so that you are sure they work fine. 

```ts
import { NextResponse, type NextRequest } from "next/server";

const isMobile = (userAgent: string) =>
  /iPhone|iPad|iPod|Android/i.test(userAgent);

const root = "/football/livescore";

export function middleware(req: NextRequest) {
  const { pathname } = req.nextUrl;

  if (
    pathname.startsWith("/api") || //  exclude all API routes
    pathname.startsWith("/static") || // exclude static files
    pathname.includes(".") // exclude all files in the public folder
  )
    return NextResponse.next();

  const userAgent = req.ua?.ua || "";

  if (pathname.startsWith(root)) {
    if (isMobile(userAgent)) {
      const league = pathname.replace(root, "");
      // league can be an empty string
      req.nextUrl.pathname = root + "/mobile" + league;
      return NextResponse.rewrite(req.nextUrl);
    }

    return NextResponse.next();
  }

  return NextResponse.next();
}
```

Not sure I've covered all of your use cases, but the above, with some tweaks, should work I believe. Careful with trailing slashes.

**Reply by andriiradkevych** [reply]
The reason why I wanted to make it this way , or by cookie , to prevent chaning of the path , I don't want to make any redirects or rewrites . By cookie , it works , but only on second reload , I don't know why ... also  I wanted to make it by `header` , but seems it doesn't wrok.. I wanted to make it by by adding ?device=mobile , to url , when it is mobile , but I'm not sure that it will not make any issues with seo ..

**Reply by icyJoseph** [reply]
> , I don't want to make any redirects or rewrites . 

But you are trying to do rewrites anyway? Rewrites keep the URL for the user, so they'll not see the mobile bit, that's internal.

And again the issue is that the request is not mutable, that's why you gotta reload if you use cookies, and why it is not there with headers.

The way I show above, does exactly the same you wanted, you are changing the path, but internally, when doing rewrites, the user will never see the mobile bit. In fact, the way you propose, users could try to reach it directly, but with middleware, you can prevent direct access to the mobile pages.

**Reply by andriiradkevych** [reply]
yes, this was the time , when I was trying whatever it has (next js ) , to make it just work , `rewrites` do job well )
